### PR TITLE
[PoC] Add KFold - External Loop.

### LINKED ISSUE #839: Cross validation feature
## 🚀 Feature
Cross-Validation is a crucial model validation techniques for assessing how the model generalizes on new data. 

### Motivation

Research papers usually require cross-validation. From my point of view, this kind of feature would simplify the work of researches.

### Pitch

I want to pass a parameter to the Trainer object to specify that I want to train the model on K-folds.

 In the case that nobody wants to make a PR, I can start working on that.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #839 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/loops_customisation/k_fold.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        return DataLoader(RandomDataset(32, 64))
+
+    def test_dataloader(self):
+        return DataLoader(RandomDataset(32, 64))
+
+    def predict_dataloader(self):
+        return DataLoader(RandomDataset(32, 64))
+
+
+class KFoldLoop(ExternalLoop):
```
Comment (SeanNaren) [inline pl_examples/loops_customisation/k_fold.py:82]: Very exciting!!

File: pytorch_lightning/utilities/boring_model.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
+# Copyright The PyTorch Lightning team.
```
Comment (tchaton) [inline pytorch_lightning/utilities/boring_model.py:1]: note for reviewing. Adding boring_model.py to utilities as it is pretty fundamental and should be part of the codebase for debugging purposes.

File: pl_examples/loops_customisation/k_fold.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        if np.max(indices) >= len(dataset) or np.min(indices) < 0:
+            raise MisconfigurationException(f"`indices` should be within [0, {len(dataset) -1}].")
+
+        self.dataset = dataset
+        self.indices = indices
+
+    def __getattr__(self, key: str):
+        if key not in self._INTERNAL_KEYS:
+            return self.dataset.__getattribute__(key)
```
Comment (justusschock) [inline pl_examples/loops_customisation/k_fold.py:70]:
```suggestion
            return getattr(self.dataset, key)
```

File: pl_examples/loops_customisation/k_fold.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+        if use_duplicated_indices:
+            indices = list(indices)
+        else:
+            indices = list(np.unique(indices))
+
+        if np.max(indices) >= len(dataset) or np.min(indices) < 0:
+            raise MisconfigurationException(f"`indices` should be within [0, {len(dataset) -1}].")
+
+        self.dataset = dataset
```
Comment (justusschock) [inline pl_examples/loops_customisation/k_fold.py:65]: maybe also do a typing.cast of self here? I usually do a typecheck if i need to do some additional wrapping.

File: pl_examples/loops_customisation/k_fold.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    @rank_zero_only
+    def on_fold_start(self, trainer, pl_module, counter):
+        log.info(f"Starting to train on fold {counter}")
+
+
+loop = KFoldLoop(5)
+model = BoringModel()
+datamodule = BoringDataModule()
+trainer = Trainer(callbacks=KFoldCallback())
+trainer.run_loop(model, datamodule=datamodule, loop=loop)
```
Comment (justusschock) [inline pl_examples/loops_customisation/k_fold.py:162]: I don't think we should introduce something like this. IMO the w should be the atomic element here and not call something that iterates on it. This should ratjer be implemented as a standalone utility class (as discussed in the corresponding issue (can't fond the number right now))

File: pytorch_lightning/loops/base.py [hunk 6]
```diff
@@ -238,3 +240,70 @@ def _load_from_state_dict(
 
         self.on_load_checkpoint(state_dict[prefix + "state_dict"])
         self.restarting = True
+
+
+class ExternalLoop(Loop):
```
Comment (justusschock) [inline pytorch_lightning/loops/base.py:245]: as written above, i really think the loops should only deal with stuff within the trainer calls not external. everything that happens outside of the trainer class should be handled by utility classes (if at all) or left to the user/flash.

**Review by justusschock — COMMENTED** [review]
TBH, I don't think this is the right way to approach this. This isn't what loops are meant for and this weakens the call hierarchy in a way we shouldn't allow. In the related issue this was discussed and approved from all sides to be a standalone class

File: pl_examples/loops_customisation/k_fold.py (hunk 5 above)
Comment (Borda) [reply pl_examples/loops_customisation/k_fold.py:162]: I guess it is just remaining after debugging

**Comment by Borda** [comment]
can we get a bigger picture of the PR?

File: pytorch_lightning/loops/base.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/loops/base.py:245]: I agree with @justusschock . There's also no bound on what options we provide here, and going in this direction will end up exposing all of the trainer internals to users, so i'm not in favor of this approach

**Comment by tchaton** [comment]
> TBH, I don't think this is the right way to approach this. This isn't what loops are meant for and this weakens the call hierarchy in a way we shouldn't allow. In the related issue this was discussed and approved from all sides to be a standalone class

Hey @justusschock.

I strongly disagree there. I believe this is what Loop are meant from the beginning.

Users should have the choice to either built on top of Lightning `fit`, `validate`, `test`, `predict` primitives + some carefully selected utilities provided by Lightning to fully control their fitting flow or switch any existing Loop within Lightning (which is definitely much advanced).

The first approach should be used when multiple `trainer.fit` are sequentially used such as in Task based learning, Cross Validation, etc... This should reduce boilerplate while enabling users to quickly experiment with new ideas or built their own framework on top of Lightning with their own callbacks system + a fully encapsulated logic.

Furthermore, if users properly implements the ExternalLoop contract, Lightning can add checkpointing + fault tolerant to their loops while maintaining full customization.

@awaelchli @williamFalcon Any thoughts there ?
Best,
T.C

File: pytorch_lightning/utilities/boring_model.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/boring_model.py:1]: I don't think the same. We already have a boring model for our tests and a boring model for bug reports and debugging. I vote for not including it in utilities.

File: pl_examples/loops_customisation/k_fold.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.loops.base import ExternalLoop
+from pytorch_lightning.utilities import rank_zero_only
+from pytorch_lightning.utilities.boring_model import BoringDataModule, BoringModel
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+seed_everything(42)
+
+
+class SplitDataset(Dataset):
```
Comment (awaelchli) [inline pl_examples/loops_customisation/k_fold.py:38]: what's the difference between this and `torch.utils.data.Subset`?
Can we just replace it as it seems to add no extra functionality

File: pl_examples/loops_customisation/k_fold.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        if name in self._INTERNAL_KEYS:
+            self.__dict__[name] = value
+        else:
+            setattr(self.dataset, name, value)
+
+    def __getitem__(self, index: int) -> Any:
+        return self.dataset[self.indices[index]]
+
+    def __len__(self) -> int:
+        return len(self.indices) - 1
```
Comment (awaelchli) [inline pl_examples/loops_customisation/k_fold.py:83]: why -1?

File: pytorch_lightning/loops/base.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/loops/base.py:245]: In the case of this kfold loop, do we need this class at all? the extra properties it defines could very well just live in the kfold loop itself as they are specific to the need there.

File: pytorch_lightning/loops/base.py [hunk 9]
```diff
@@ -238,3 +240,70 @@ def _load_from_state_dict(
...
+    """This Loop is meant wrap trainer calls"""
+
+    @property
+    def trainer(self) -> Optional["pl.Trainer"]:
+        return self._trainer
+
+    @trainer.setter
+    def trainer(self, trainer: "pl.Trainer"):
+        """Connects this loop's trainer and its children"""
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:254]: this already exists in the base loop :)

File: pytorch_lightning/trainer/data_loading.py [hunk 10]
```diff
@@ -252,6 +252,67 @@ def replace_sampler(self, dataloader: DataLoader, sampler, mode: Optional[Runnin
         dataloader = dl_cls(**dl_kwargs)
         return dataloader
 
+    def apply_user_function(self, dataloader: DataLoader, user_function: Callable) -> DataLoader:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:255]: what if we had instead:
```python

def train_dataset(self):
    return Dataset(...)
    
def create_dataloader(self, dataset):
     return DataLoader(dataset, ...)
```
     
then by default we can call `module.create_dataloader(module.train_dataset)` and if needed one can intercept the dataset before passing it to the dataloader creation.

File: pl_examples/loops_customisation/k_fold.py (hunk 7 above)
Comment (tchaton) [reply pl_examples/loops_customisation/k_fold.py:38]: I didn't know about it :) Thanks.

File: pl_examples/loops_customisation/k_fold.py (hunk 8 above)
Comment (tchaton) [reply pl_examples/loops_customisation/k_fold.py:83]: Removed !

File: pytorch_lightning/loops/base.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:254]: Removed !

File: pytorch_lightning/trainer/data_loading.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:255]: Done !

File: pytorch_lightning/utilities/boring_model.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/boring_model.py:1]: @carmocca Any thoughts ?

Comment (carmocca) [reply pytorch_lightning/utilities/boring_model.py:1]: So I ran into the same issue in https://github.com/PyTorchLightning/pytorch-lightning/pull/7614#discussion_r671183652 (blocked for the same reason).

Some `pl_examples` rely on the `MNISTDataModule`, which is in the test directory.

We currently include everything in our distribution, which is what I try to avoid in the linked PR.

However, CI fails because then the pl_examples do not have access to the `MNISTDataModule` implementation if we do.

So we have two real options:

1. Duplicate the `BoringModel`/`MNISTDataModule`/ implementations in the `pl_examples` directory
2. Move `BoringModel`/`MNISTDataModule` to the source tree (what this PR does) to avoid code duplication.

I think I prefer (2). If you guys agree I can do this change in a separate PR

cc @Borda

Comment (carmocca) [reply pytorch_lightning/utilities/boring_model.py:1]: Lets continue discussing in https://github.com/PyTorchLightning/pytorch-lightning/pull/8776

File: pl_examples/loops_customisation/k_fold.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        return self.trainer.fit(self.trainer.lightning_module, datamodule=self.trainer.datamodule)
+
+    def on_advance_end(self) -> None:
+        self.current_fold += 1
+        # stored best weight path for this fold
+        self.best_model_paths.append(self.trainer.checkpoint_callback.best_model_path)
+
+    # utilities for creating a hold
+    def process_dataset(self, stage: str, dataset: Dataset) -> Subset:
+        kfold = KFold(self.num_folds, random_state=42, shuffle=True)
```
Comment (justusschock) [inline pl_examples/loops_customisation/k_fold.py:191]: is a dependency for sklearn worth it for just this?
Should we maybe have a more general abstract function `create_splits` the user has to implement? Since there are so many different ways to create data splits.

File: pl_examples/loops_customisation/k_fold.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+from torch.utils.data.dataset import Dataset, Subset
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning import LightningDataModule, seed_everything
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.loops.external_loop import ExternalLoop
+from pytorch_lightning.utilities import rank_zero_only
+from pytorch_lightning.utilities.boring_model import BoringModel, RandomDataset
+
+seed_everything(42)
```
Comment (justusschock) [inline pl_examples/loops_customisation/k_fold.py:34]:
```suggestion
```
rather not seed anything globally

File: pl_examples/loops_customisation/k_fold.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+        self.restarting = restarting
+
+    @staticmethod
+    def loop_base_callback() -> Type[Callback]:
+        class BaseKFoldCallback(Callback):
+            @rank_zero_only
+            def on_fold_start(self, trainer, pl_module, counter):
+                """Override with your own logic"""
+
+        return BaseKFoldCallback
```
Comment (justusschock) [inline pl_examples/loops_customisation/k_fold.py:154]: can't we define this outside this class but in the file namespace?

File: pl_examples/loops_customisation/k_fold.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+    @rank_zero_only
+    def on_fold_start(self, trainer, pl_module, counter) -> None:
+        log.info(f"Starting to train on fold {counter}")
+
+
+loop = KFoldLoop(5)
+model = BoringModel()
+datamodule = BoringDataModule()
+loop.connect_trainer(max_epochs=10, callbacks=KFoldCallback())
```
Comment (awaelchli) [inline pl_examples/loops_customisation/k_fold.py:215]: alternatively these could be passed in through init via an argument trainer_kwargs.

File: pl_examples/loops_customisation/k_fold.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+from pytorch_lightning.utilities import rank_zero_only
+from pytorch_lightning.utilities.boring_model import BoringModel, RandomDataset
+
+seed_everything(42)
+
+
+class BaseDataModule(LightningDataModule):
+    def __init__(self):
+        super().__init__()
+        self.non_picklable = None
```
Comment (awaelchli) [inline pl_examples/loops_customisation/k_fold.py:40]: curious, what was the idea here? seems left over :)

File: pl_examples/loops_customisation/k_fold.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        self.checkpoint_state: Optional[str] = None
```
Comment (awaelchli) [inline pl_examples/loops_customisation/k_fold.py:41]: same question here :)

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 17]
```diff
@@ -471,9 +476,13 @@ def _get_lightning_module_state_dict(self) -> Dict[str, torch.Tensor]:
...
-        return {
+        state_dict = {
             "fit_loop": self.trainer.fit_loop.state_dict(),
             "validate_loop": self.trainer.validate_loop.state_dict(),
             "test_loop": self.trainer.test_loop.state_dict(),
             "predict_loop": self.trainer.predict_loop.state_dict(),
         }
+        external_loop = getattr(self.trainer, "external_loop", None)
+        if external_loop:
+            state_dict.update({"external_loop": external_loop.state_dict()})
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:487]: can there be more than one external loop. I mean, one external loop nested inside another?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 18]
```diff
@@ -471,9 +476,13 @@ def _get_lightning_module_state_dict(self) -> Dict[str, torch.Tensor]:
...
 
     def _get_loops_state_dict(self) -> Dict[str, Any]:
-        return {
+        state_dict = {
             "fit_loop": self.trainer.fit_loop.state_dict(),
             "validate_loop": self.trainer.validate_loop.state_dict(),
             "test_loop": self.trainer.test_loop.state_dict(),
             "predict_loop": self.trainer.predict_loop.state_dict(),
         }
+        external_loop = getattr(self.trainer, "external_loop", None)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:485]: perhaps trainer can have a property like one we have for the other loops?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 19]
```diff
@@ -186,7 +186,7 @@ def restore_callbacks(self) -> None:
             )
         self.trainer.on_load_checkpoint(self._loaded_checkpoint)
 
-    def restore_loops(self) -> None:
+    def restore_loops(self, restore_external_loop: bool = False) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:189]: it's probably enough if this is controlled below by the existence of an external loop as checked below.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:487]: It could and Loop will automatically gather their children states.

**Comment by turian** [comment]
@tchaton will this PR make this thru?

**Comment by tchaton** [comment]
Hey @turian,

I am quite unsure. I see some advantages for an External Loop, but I believe it is maybe too much of a learning curve for new users.
What are your thoughts there ?

Best,
T.C

**Comment by turian** [comment]
@tchaton where is the quickstart version of how to use this?

**Comment by tchaton** [comment]
Hey @turian,

Are you interested in ExternalLoop or KFold ?

Best,
T.C

**Comment by turian** [comment]
@tchaton KFold. What are the use-cases for external loop?
